components/raftstore-v2/src/router/response_channel.rs

```diff
@@ -0,0 +1,448 @@
+/// have no payload. At most 30 states can be defined. There can be only one
+/// type of payload.
+struct EventCore<Res> {
+    /// Every event will have two bits.
+    /// - 0b00 means the event is not fired and not watched.
+    /// - 0b01 means the event is fired and not watched.
+    /// - 0b10 means the event is not fired and watched.
+    /// - 0b11 means the event is fired and watched.
+    event: AtomicU64,
+    res: UnsafeCell<Option<Res>>,
+    // Waker can be changed, need to use `AtomicWaker` to guarantee no data race.
+    waker: AtomicWaker,
+}
+
+unsafe impl<Res: Send> Send for EventCore<Res> {}
+
+const PAYLOAD_EVENT: u64 = 0;
+const CANCEL_EVENT: u64 = 31;
+
+#[inline]
+const fn subscribed_bit_of(event: u64) -> u64 {
+    1 << (event * 2)
+}
+
+#[inline]
+const fn set_bit_of(event: u64) -> u64 {
+    1 << (event * 2 + 1)
+}
+
+impl<Res> EventCore<Res> {
+    #[inline]
+    pub fn notify_event(&self, event: u64) {
+        let previous = self.event.fetch_or(set_bit_of(event), Ordering::AcqRel);
+        if previous & subscribed_bit_of(event) != 0 {
+            self.waker.wake()
+        }
+    }
+
+    /// Set the result.
+    ///
+    /// After this call, no events should be notified.
+    #[inline]
+    pub fn set_result(&self, result: Res) {
+        unsafe {
+            *self.res.get() = Some(result);
+        }
+        let previous = self.event.fetch_or(
+            set_bit_of(PAYLOAD_EVENT) | set_bit_of(CANCEL_EVENT),
+            Ordering::AcqRel,
+        );
+        if previous & subscribed_bit_of(PAYLOAD_EVENT) != 0 {
+            self.waker.wake()
+        }
+    }
+
+    /// Cancel all subscribers.
+    ///
+    /// After this call, no events should be notified and no result should be
+    /// set.
+    #[inline]
+    pub fn cancel(&self) {
+        let mut previous = self
+            .event
+            .fetch_or(set_bit_of(CANCEL_EVENT), Ordering::AcqRel);
+        let subscribed_bit = subscribed_bit_of(0);
+        while previous != 0 {
+            // Not notified yet.
+            if previous & 0b11 == subscribed_bit {
+                self.waker.wake();
+                return;
+            }
+            previous >>= 2;
+        }
+    }
+}
+
+struct WaitEvent<'a, Res> {
+    event: u64,
+    core: &'a EventCore<Res>,
+}
+
+#[inline]
+fn check_bit(e: u64, set_bit: u64) -> Option<bool> {
+    if e & set_bit != 0 {
+        return Some(true);
+    }
+    let cancel_bit = set_bit_of(CANCEL_EVENT);
+    if e & cancel_bit != 0 {
+        return Some(false);
+    }
+    None
+}
+
+impl<'a, Res> Future for WaitEvent<'a, Res> {
+    type Output = bool;
+
+    #[inline]
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let event = &self.core.event;
+        let mut e = event.load(Ordering::Relaxed);
+        let set_bit = set_bit_of(self.event);
+        if let Some(b) = check_bit(e, set_bit) {
+            return Poll::Ready(b);
+        }
```

Comment:
if the u64 event is enum,  set_bit_of can be simply renamed as check_fired(&self).
now the usage of this is weird---first i need to atomic get the event value e, then get its set bit, then call check_bit(). It can be wrapped into one func with very clear semantic. 
Same logic repeats in WaitResult

Reply:
`check_bit` is also used in other places. Wrapping them into a function just to avoid repeating bit operations, which is very easy to make mistake.